command/agentproxyshared/cache/static_secret_cache_updater.go

```diff
@@ -0,0 +1,372 @@
+type StaticSecretCacheUpdater struct {
+	client     *api.Client
+	leaseCache *LeaseCache
+	logger     hclog.Logger
+	tokenSink  sink.Sink
+}
+
+// StaticSecretCacheUpdaterConfig is the configuration for initializing a new
+// StaticSecretCacheUpdater.
+type StaticSecretCacheUpdaterConfig struct {
+	Client     *api.Client
+	LeaseCache *LeaseCache
+	Logger     hclog.Logger
+	// TokenSink is a token sync that will have the latest
+	// token from auto-auth in it, to be used in event system
+	// connections.
+	TokenSink sink.Sink
+}
+
+// NewStaticSecretCacheUpdater creates a new instance of a StaticSecretCacheUpdater.
+func NewStaticSecretCacheUpdater(conf *StaticSecretCacheUpdaterConfig) (*StaticSecretCacheUpdater, error) {
+	if conf == nil {
+		return nil, errors.New("nil configuration provided")
+	}
+
+	if conf.LeaseCache == nil || conf.Logger == nil {
+		return nil, fmt.Errorf("missing configuration required params: %v", conf)
+	}
+
+	if conf.Client == nil {
+		return nil, fmt.Errorf("nil API client")
+	}
+
+	if conf.TokenSink == nil {
+		return nil, fmt.Errorf("nil token sink")
+	}
+
+	return &StaticSecretCacheUpdater{
+		client:     conf.Client,
+		leaseCache: conf.LeaseCache,
+		logger:     conf.Logger,
+		tokenSink:  conf.TokenSink,
+	}, nil
+}
+
+// streamStaticSecretEvents streams static secret events and updates
+// the cache when updates are notified. This method will return errors in cases
+// of failed updates, malformed events, and other.
+// For best results, the caller of this function should retry on error with backoff,
+// if it is desired for the cache to always remain up to date.
+func (updater *StaticSecretCacheUpdater) streamStaticSecretEvents(ctx context.Context) error {
+	// First, ensure our token is up-to-date:
+	updater.client.SetToken(updater.tokenSink.(sink.SinkReader).Token())
+	conn, err := updater.openWebSocketConnection(ctx)
+	if err != nil {
+		return fmt.Errorf("error when opening event stream: %w", err)
+	}
+	defer conn.Close(websocket.StatusNormalClosure, "")
+
+	// before we check for events, update all of our cached
+	// kv secrets, in case we missed any events
+	// TODO: to be implemented in a future PR
+
+	for {
+		_, message, err := conn.Read(ctx)
+		if err != nil {
+			// The caller of this function should make the decision on if to retry. If it does, then
+			// the websocket connection will be retried, and we will check for missed events.
+			return fmt.Errorf("error when attempting to read from event stream, reopening websocket: %w", err)
+		}
+		updater.logger.Trace("received event", "message", string(message))
+		messageMap := make(map[string]interface{})
+		err = json.Unmarshal(message, &messageMap)
+		if err != nil {
+			return fmt.Errorf("error when unmarshaling event, message: %s\nerror: %w", string(message), err)
+		}
+		data, ok := messageMap["data"].(map[string]interface{})
+		if !ok {
+			return fmt.Errorf("unexpected event format, message: %s\nerror: %w", string(message), err)
```

Comment:
I think the existing message is clearer, as while this is about a specific part of the event, ultimately, something has gone very wrong if the event is in the wrong shape. In particular, we'd get this message if there was no `data` element in the JSON, so data format isn't wrong, it's not present :)

Reply:
To put it another way, if we receive `{ "foo": "bar"}` as our event, the event format is wrong, as opposed to the format of the (missing) `data` object. There are no valid events without the `data` object, so something has gone quite wrong with the event if this ever happens.

I believe this applies to the other, similar comments you had too, but let me know if you think it doesn't!

Reply:
What worried me was that we have four identical error message strings in the code which might make pinpointing the exact code path when debugging more difficult, but I guess maybe in that case the value of `message` could point us in the right direction? 

In general, I like when error message strings are as unique as possible to allow searching the codebase during incidents, etc. simpler, but it might be an abundance of caution from my side 😄

Reply:
It should (hopefully) be immediately obvious from the error what's wrong, but you're right, knowing where the error came from could definitely help! Let me update to e.g. `unexpected event format when decoding 'data' element`, does that work? :)

Reply:
Yeah, that sounds spot on. Thanks, Violet!

Reply:
Updated! Thanks a bunch! :)